src/modules/kvs/commit.c

```diff
@@ -242,24 +242,34 @@ static int store_cache (commit_t *c, int current_epoch, json_t *o,
  */
 static int commit_unroll (commit_t *c, int current_epoch, json_t *dir)
 {
-    json_t *value;
-    json_t *subdir, *key_value;
-    json_t *tmpdirent;
+    json_t *dir_entry;
+    json_t *dir_data;
+    json_t *tmp;
     href_t ref;
     int ret;
     struct cache_entry *hp;
-    void *iter = json_object_iter (dir);
+    void *iter;
+
+    if (!treeobj_is_dir (dir)) {
+        errno = EPERM;
```

Comment:
Seems like this is an internal "can't happen" scenario, so EPERM to the user is questionable?   Maybe this should be an assert?

Tracing back in `commit_process()`, it looks like "rootcpy" is created with `treeobj_copy_dir()` which would fail  with EINVAL if root was not a directory (another problem there: the errno is overwritten with ENOMEM).  I'm hoping there's futher protection that would prevent root from being set to anything other than a directory, and yes that seems to be true:
```
$ flux kvs put .=42
2017-08-24T16:08:50.437304Z kvs.err[0]: commit failed: Invalid argument
flux-kvs: flux_kvs_commit: Invalid argument
```
(um, not sure about that causing a LOG_ERR message)

But anyway, an assertion should be OK here.